### DISCUSSION TITLE
Unable to modify a file in parallel if it contains links

### DISCUSSION THREAD

**Discussion by flotus1** [CGNS/CGNS#300] [CGNStalk]
Following the suggestion here https://github.com/CGNS/CGNS/discussions/297 I am trying to write solution files that contain links to a single mesh file. Worked like a charm in serial, but I can't wrap my head around the errors I am getting when trying to do the same in parallel.

After writing the mesh file, I create a solution file with all the metadata that's needed, and links to the mesh file. The goal is to write various results to that same file at a later point. Same procedure I used in a serial version without any issues. Problem is: after closing the file, I can not open it again later with cgp_open_f in "CG_MODE_MODIFY" mode. I get a rather generic MPI error here:
```
call cgp_open_f(trim(adjustl(filename)), CG_MODE_MODIFY, cgns_index_file, ier)
if (ier .NE. CG_OK) call cgp_error_exit_f
```
```
*** An error occurred in MPI_File_open
*** reported by process [2624716801,2]
*** on communicator MPI_COMM_WORLD
*** MPI_ERR_COMM: invalid communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
2 more processes have sent help message help-mpi-errors.txt / mpi_errors_are_fatal
```
MPI_COMM_WORLD is the correct communicator. I can open the file in serial mode with cg_open_f and CG_MODE_MODIFY no problem. CGNSCHECK finds no errors in the file. Looking at it with CGNSVIEW everything is as it should be. All nodes are there and the links are valid. I can even open the file in ParaView and look at the mesh that was linked. Error return values from all cgns calls are zero. I just can't seem to open is again to modify with cgp_open_f to write the actual data.
I get the same error if I try to create the file with the serial or parallel cgns calls first.
If I comment out the part that writes the links to the file, I can open it with cgp_open_f. But then of course the links to the mesh are missing.
```
! Create a link to the grid coordinates in the grid file
call cg_goto_f(cgns_index_file, cgns_index_base, ier, 'Zone_t', cgns_index_zone, 'end')
if (ier .NE. CG_OK) call cg_error_exit_f
link_path = 'Base/Zone 1/GridCoordinates'
call cg_link_write_f('GridCoordinates', cgns_mesh_name, trim(adjustl(link_path)),ier)
if (ier .NE. CG_OK) call cg_error_exit_f
! Create a link to the element connectivity in the grid file
link_path = 'Base/Zone 1/Elem'
call cg_link_write_f('Elem', cgns_mesh_name, trim(adjustl(link_path)), ier)
if (ier .NE. CG_OK) call cg_error_exit_f
```
Pretty stuck here, what else can I try?

**Comment by flotus1** [comment]
So far, I came up with 2 different workarounds for this bug. Let's call them A and B

**Implementation A** prepares the solution files as usual, but leaves out the links to the mesh that were causing the problem. Then the solution data is written to the file. And in the end after all parallel writes are done, a single thread adds the mesh links to the solution file. This seems to work ok as far as correctness is concerned. CGNSCHECK finds no errors with the files, and I can load them in ParaView no problem. Performance looks like this:
![cgns_parallel_timing_a](https://user-images.githubusercontent.com/57221355/133222758-46f7b37a-98cc-4569-be37-bcf1854c88ff.png)
![cgns_parallel_writespeed_a](https://user-images.githubusercontent.com/57221355/133222774-8f77abd4-ccad-493d-ae52-acfe749ed9d9.png)

**Implementation B** gets around the issue by simply not closing the solution file until the end of a time step. For correctness, it seems to be the same as A. Here is what that looks like in terms of performance:
![cgns_parallel_timing_b](https://user-images.githubusercontent.com/57221355/133223219-4e924272-b1da-4668-92c4-c2ee30264dfc.png)
![cgns_parallel_writespeed_b](https://user-images.githubusercontent.com/57221355/133223232-f2ae1a49-e802-47db-97e0-268ff99a1386.png)

So implementation B is clearly superior in terms of performance. Any comments on the approaches I picked so far?
What worries me a lot is the performance trend over the number of threads. It goes down rather fast. This was tested on my local workstation with 64 cores. But what I am developing here needs to run on thousands of cores later. What's your solution to that? Collect the results from all MPI threads to a lower number of "I/O-threads" that handle writing the results?

By the way, my performance measurements above take into account everything that needs to happen in order to get a CGNS file at the end. From opening the file, creating metadata, writing the data, and lastly closing the file. The example in src/ptests/pcgns_ftest.F90 hides all that overhead by only measuring the time it takes to write the raw data.

**Reply by brtnfld** [reply]
Are you writing to a parallel file system? I would expect the behavior to be different if you are.

If the CGNS file structure is the issue, can you create the CGNS file with one rank, close the file and open the file by all the ranks to write the raw data? 

If you have a standalone test program available,  we could look into profiling the program and see where the time is being spent.

**Reply by flotus1** [reply]
I don't think that I am using a parallel file system. It's just a regular old NVMe SSD with an ext4 file system on it.
I can't create the solution files serial. As stated in the initial post, it leads to the same problem when opening it in parallel afterwards.
If you want to take a look, standalone f90 source files are attached below. Had to change the file extension to .txt to upload them here. Parameters for number of cells (ni, nj, nk) and time steps (nsteps) are hard-coded for now. You can find them towards the bottom of the files. CGNS is assumed to be built with 64-bit.
[cgns_parallel_implementation_a.txt](https://github.com/CGNS/CGNS/files/7163708/cgns_parallel_implementation_a.txt)
[cgns_parallel_implementation_b.txt](https://github.com/CGNS/CGNS/files/7163711/cgns_parallel_implementation_b.txt)

**Comment by brtnfld** [comment]
Thanks, can you also enter a JIRA issue for the external links error (with reproducer) so we can keep track of it. You can also enter other issues that you would like to see looked into.  Otherwise, it will probably not be scheduled to be investigated. 
https://cgnsorg.atlassian.net

**Reply by flotus1** [reply]
I'm pretty reluctant to file a bug report on anything at this point, because I really don't know what I'm doing with CGNS. So chances are high that I am just using it wrong, and someone with more experience can just tell me what to do differently. I will keep that in mind for later though.

**Comment by NovA80** [comment]
Seems it's related with a long standing issue with links in serial files as well: https://cgnsorg.atlassian.net/browse/CGNS-86. BTW, that issue mentions the mailist message, I'd reported a while ago. And this is probably an HDF bug, so no soon solving is expected :)

**Reply by brtnfld** [reply]
The issue with CGNS-86 is this:

(1) CGNS opens the file to which the link points as read-only.
(2) CGNS now opens the parent file as read-write.
      (2a) CGNS now tries to open the external link.
      (2b) HDF5 tries to open the file pointed to by the external link as read-write (this is inherited from the parent file)
      (2c) HDF5 fails because the two access modes are not compatible.

Step 2b is probably not apparent with CGNS that this is going on as the details are HDF5 specific.

There is a chart in HDF5 (H5Fint.c) that explains this:
```
* The following two tables show results of file opens for single and concurrent access:
 *
 * SINGLE PROCESS ACCESS                        CONCURRENT ACCESS
 *
 *             #1st open#                                   #1st open#
 *             -- SR SR -- -- SR SR --                      -- SR SR -- -- SR SR --
 *             -- -- SW SW SW SW -- --                      -- -- SW SW SW SW -- --
 *              W  W  W  W  R  R  R  R                       W  W  W  W  R  R  R  R
 * #2nd open#                                   #2nd open#
 *            --------------------------                   --------------------------
 *   -- --  W | s  x  x  s  x  x  f  f |          -- --  W | f  x  x  f  x  x  f  f |
 *   SR --  W | x  x  x  x  x  x  x  x |          SR --  W | x  x  x  x  x  x  x  x |
 *   SR SW  W | x  x  x  x  x  x  x  x |          SR SW  W | x  x  x  x  x  x  x  x |
 *   -- SW  W | f  x  x  s  x  x  f  f |          -- SW  W | f  x  x  f  x  x  f  f |
 *   -- SW  R | x  x  x  x  x  x  x  x |          -- SW  R | x  x  x  x  x  x  x  x |
 *   SR SW  R | x  x  x  x  x  x  x  x |          SR SW  R | x  x  x  x  x  x  x  x |
 *   SR --  R | s  x  x  s  x  x  s  f |          SR --  R | f  x  x  s  x  x  s  s |
 *   -- --  R | s  x  x  s  x  x  s  s |          -- --  R | f  x  x  f  x  x  s  s |
 *            --------------------------                   --------------------------
 *
 *      Notations:
 *        W:  H5F_ACC_RDWR
 *        R:  H5F_ACC_RDONLY
 *        SW: H5F_ACC_SWMR_WRITE
 *        SR: H5F_ACC_SWMR_READ
 *
 *        x: the first open or second open itself fails due to invalid flags combination
 *        f: the open fails with flags combination from both the first and second opens
 *        s: the open succeeds with flags combination from both the first and second opens
 *
```

The error can be reproduced in pure HDF5 by doing:
```
file = H5Fopen ("zone1.cgns", H5F_ACC_RDONLY, H5P_DEFAULT);
file2 = H5Fopen ("links.cgns", H5F_ACC_RDWR, H5P_DEFAULT);
 gid = H5Gopen(file2,"/Base/Zone1/ link", H5P_DEFAULT);
```

So this is a CGNS user error. HDF5 is designed to have this behavior. The HDF5 returned error message is not the best.

**Reply by NovA80** [reply]
As for CGNS-86, I have different workflow that triggers the issue:
(1) CGNS opens the parent file as read-only.
(1a) CGNS now opens the external link as read-only.
(2) CGNS closes the parent file
(2a) CGNS *not* closes the external link, and it remains openned read-only.
(3) CGNS re-opens the parent file as read-write
(3a) CGNS now tries to open the external link as read-write.
(3b) HDF5 tries to open the file pointed to by the external link as read-write (this is inherited from the parent file)
(3c) HDF5 fails because the linked to file was not closed in step (2a), and the two access modes are not compatible.

The issue is in step (2a). I would like to workaround it by manually closing the linked file, but it seems there is no direct access to the corresponding file descriptor.

**Reply by NovA80** [reply]
Actually under Unix, this problem of not closing the pointed to file is not a big deal. Just use read-write mode in step (1) and in step (3a) external file will be reopened silently. But under Windows such reopening fails.

**Reply by brtnfld** [reply]
Do you mean in step 1a that you are not using cg_open to open the external link file?

**Reply by brtnfld** [reply]
Sigh, that is the problem of CGNS opening HDF5 objects in those externally linked files and not closing them. When H5Fclose is called on the parent file, it can't close those external files because the behavior in HDF5 is to delay closing those external files until all the objects in them are closed. I need to look into closing those objects in CGNS. One can force HDF5 to close all open objects and close the file, but it isn't easy to set that property on an external file for which HDF5 opens. At least I've not gotten it to work yet.

**Reply by NovA80** [reply]
> Do you mean in step 1a that you are not using cg_open to open the external link file?

Yes, I've meant exactly that. A user initiates step (1) with 'cg_open()', and step (1a) is done by CGNS itself. Steps (2) & (3) are also started by user using ’cg_close()’ & ’cg_open()’ respectively.

**Reply by NovA80** [reply]
> Sigh, that is the problem of CGNS opening HDF5 objects in those externally linked files and not closing them.

BTW, it looks like accessing an object in the linked file, unconditionally opens that file even if it is already openned. For example, consider a field file with 1000 zones, each with a Coordinate{X, Y, Z} linked to a data in a corresponding zone of the external file. When one cg_open() such a file and touches the Coordinate arrays, the linked to file reopens for each zone. This may take ages for complex grids.

**Reply by brtnfld** [reply]
We could increase the number of files held open in HDF5 from external link access by using H5Pset_elink_file_cache_size. Do you have an example that I can test?

**Reply by brtnfld** [reply]
> As for CGNS-86, I have different workflow that triggers the issue:
> (1) CGNS opens the parent file as read-only.
> (1a) CGNS now opens the external link as read-only.
> (2) CGNS closes the parent file
> (2a) CGNS _not_ closes the external link, and it remains openned read-only.
> (3) CGNS re-opens the parent file as read-write
> (3a) CGNS now tries to open the external link as read-write.
> (3b) HDF5 tries to open the file pointed to by the external link as read-write (this is inherited from the parent file)
> (3c) HDF5 fails because the linked to file was not closed in step (2a), and the two access modes are not compatible.
> 
> The issue is in step (2a). I would like to workaround it by manually closing the linked file, but it seems there is no direct access to the corresponding file descriptor.

After some investigation, this is related to

https://cgnsorg.atlassian.net/browse/CGNS-237

I was able to get the test to pass by using H5Iiterate and closing open groups that are missed at file close. Looking into what groups are not getting closed and why. Even if I use the CGNS id tracking option they are not getting closed.
